### PR TITLE
Add jj diffs --stat option

### PR DESCRIPTION
**Pull request by ob**
This is a first cut at #2066. It needs a better way of getting the terminal width (hardcoded for now to 120) and maybe some tests... I did a few visual comparisons to `git diff` and it looks like it works...

```
jj ❯ COLUMNS=120 git diff --stat 2102de94 e414f3b7
 CHANGELOG.md                       |   4 +++
 cli/src/cli_util.rs                |  34 ++++++-----------------
 cli/src/commands/mod.rs            |  51 +++++++++++++++++-----------------
 cli/src/config-schema.json         |   5 ++++
 cli/src/config/colors.toml         |   1 +
 cli/src/diff_util.rs               |  20 ++++++++++----
 cli/src/merge_tools.rs             |  55 ++++++++++++++++++++-----------------
 cli/tests/test_describe_command.rs |  89 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-
 cli/tests/test_diff_command.rs     |  23 ++++++++++++++++
 docs/config.md                     |   4 +--
 docs/operation-log.md              |   2 +-
 lib/src/conflicts.rs               |  17 ++++--------
 lib/src/default_index_store.rs     | 115 +++++++++++++++++++++++------------------------------------------------------
 lib/src/gitignore.rs               |   7 ++---
 lib/src/local_backend.rs           |   4 +--
 lib/src/repo.rs                    |   7 +++--
 lib/src/working_copy.rs            |  76 ++++++++++++++++++++++++---------------------------
 lib/src/workspace.rs               |   7 ++---
 lib/tests/test_conflicts.rs        |   8 +-----
 19 files changed, 288 insertions(+), 241 deletions(-)
```

```
jj ❯ cargo build && ./target/debug/jj diff --stat --from 2102de94 --to e414f3b7
    Finished dev [unoptimized + debuginfo] target(s) in 0.15s
 CHANGELOG.md                       |    4 +++
 cli/src/cli_util.rs                |   34 +++++------------------
 cli/src/commands/mod.rs            |   51 +++++++++++++++++------------------
 cli/src/config/colors.toml         |    1 +
 cli/src/config-schema.json         |    5 +++
 cli/src/diff_util.rs               |   20 ++++++++++----
 cli/src/merge_tools.rs             |   55 +++++++++++++++++++++-----------------
 cli/tests/test_describe_command.rs |   89 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-
 cli/tests/test_diff_command.rs     |   23 ++++++++++++++++
 docs/config.md                     |    4 +-
 docs/operation-log.md              |    2 +-
 lib/src/conflicts.rs               |   17 ++++--------
 lib/src/default_index_store.rs     |  115 +++++++++++++++++++++++--------------------------------------------------------
 lib/src/gitignore.rs               |    7 +---
 lib/src/local_backend.rs           |    4 +--
 lib/src/repo.rs                    |    7 +++--
 lib/src/working_copy.rs            |   76 +++++++++++++++++++++++++---------------------------
 lib/src/workspace.rs               |    7 ++---
 lib/tests/test_conflicts.rs        |    8 +-----
 19 files changed, 288 insertions(+), 241 deletions(-)
```

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

File: cli/src/diff_util.rs [hunk 1]
```diff
@@ -749,6 +759,109 @@ pub fn show_diff_summary(
...
+                let stat = get_diff_stat(&repo_path, &left_content, &[]);
+                stats.push(stat);
+            }
+        }
+        max_path_length = max(max_path_length, repo_path.to_internal_file_string().len());
+        max_diffs = stats
+            .iter()
+            .map(|stat| max(max_diffs, stat.added + stat.removed))
+            .max()
+            .unwrap_or(0);
```
Comment (martinvonz) [inline cli/src/diff_util.rs:824]: If I'm reading this correctly, it's O(n^2). I suppose you meant to put it outside the loop and write it like this:

```rust
let max_diffs = stats.iter().map(|stat| stat.added + stat.removed).max().unwrap_or(0);
```

File: cli/src/diff_util.rs [hunk 2]
```diff
@@ -749,6 +759,109 @@ pub fn show_diff_summary(
...
+        for stat in &stats {
+            total_added += stat.added;
+            total_removed += stat.removed;
+            let bar_added = (stat.added as f64 * factor).round() as usize;
+            let bar_removed = (stat.removed as f64 * factor).round() as usize;
+            // pad to max_path_length
+            write!(
+                formatter.labeled("hunk-header"),
+                " {:<max_path_length$} | {:>4}",
+                workspace_command.format_file_path(&stat.path),
```
Comment (martinvonz) [inline cli/src/diff_util.rs:847]: It's probably best to save this formatted path instead of the `RepoPath` in `DiffStat`, so the math based on its length is accurate (it's a relative path, so matters if you're in a subdirectory).

**Review by martinvonz — COMMENTED** [review]
Nice! Thank you!

> and maybe some tests..

Yes, please :)

Also, please squash the second commit into the first and address comments by amending the commit. See https://github.com/martinvonz/jj/blob/main/docs/contributing.md#code-reviews.

File: cli/src/diff_util.rs (hunk 1 above)
Comment (ob) [reply cli/src/diff_util.rs:824]: lol, yeah... I changed how I wrote it a few times and ended up with both calls to `max()`... good catch....

File: cli/src/diff_util.rs [hunk 3]
```diff
@@ -749,6 +759,109 @@ pub fn show_diff_summary(
...
+    };
+
+    formatter.with_label("diff", |formatter| {
+        let mut total_added = 0;
+        let mut total_removed = 0;
+        for stat in &stats {
+            total_added += stat.added;
+            total_removed += stat.removed;
+            let bar_added = (stat.added as f64 * factor).round() as usize;
+            let bar_removed = (stat.removed as f64 * factor).round() as usize;
```
Comment (ob) [inline cli/src/diff_util.rs:842]: We should switch these two calls to `ceil()` to avoid empty bars in the case where the diff count is 1.

File: cli/src/diff_util.rs [hunk 4]
```diff
@@ -749,6 +759,109 @@ pub fn show_diff_summary(
...
+        max_path_length = max(max_path_length, repo_path.to_internal_file_string().len());
+        max_diffs = stats
+            .iter()
+            .map(|stat| max(max_diffs, stat.added + stat.removed))
+            .max()
+            .unwrap_or(0);
+    }
+
+    // TODO: Figure out how to get the terminal width and removed the hardcoded 120.
+    let max_bar_length = 120 - max_path_length - " | ".len() - max_diffs.to_string().len() - 1;
```
[the remark this answers is not in the document]
Comment (ob) [reply cli/src/diff_util.rs]: Thanks for the pointer. Done.

File: cli/src/diff_util.rs [hunk 5]
```diff
@@ -37,7 +38,7 @@ use crate::ui::Ui;
 
 #[derive(clap::Args, Clone, Debug)]
 #[command(group(clap::ArgGroup::new("short-format").args(&["summary", "types"])))]
-#[command(group(clap::ArgGroup::new("long-format").args(&["git", "color_words", "tool"])))]
+#[command(group(clap::ArgGroup::new("long-format").args(&["git", "color_words", "tool", "stat"])))]
```
[the remark this answers is not in the document]
Comment (ob) [reply cli/src/diff_util.rs]: I moved it to the "short-format" section but I suspect more changes would be needed to support combining options like this no? Even `summary` with `types` doesn't work:

```
$ jj diff --summary --types  --from=d039008d --to=78170b42
error: the argument '--summary' cannot be used with '--types'
```

File: cli/src/diff_util.rs [hunk 6]
```diff
@@ -60,11 +61,14 @@ pub struct DiffFormatArgs {
     /// Generate diff by external command
     #[arg(long)]
     pub tool: Option<String>,
+    #[arg(long)]
+    pub stat: bool,
```
[the remark this answers is not in the document]
Comment (ob) [reply cli/src/diff_util.rs]: Thanks for noticing... consistency is important IMHO. Moved.

File: cli/src/diff_util.rs [hunk 7]
```diff
@@ -749,6 +759,109 @@ pub fn show_diff_summary(
...
+    path: RepoPath,
+    added: usize,
+    removed: usize,
+}
+
+fn get_diff_stat<'content>(
+    repo_path: &RepoPath,
+    left_content: &'content [u8],
+    right_content: &'content [u8],
+) -> DiffStat {
```
[the remark this answers is not in the document]
Comment (ob) [reply cli/src/diff_util.rs]: Ah, good catch. Dropped.

File: cli/src/diff_util.rs (hunk 1 above)
Comment (ob) [reply cli/src/diff_util.rs]: Fixed.

File: cli/src/diff_util.rs (hunk 2 above)
Comment (ob) [reply cli/src/diff_util.rs]: Good point. Fixed.

File: cli/src/diff_util.rs (hunk 3 above)
Comment (ob) [reply cli/src/diff_util.rs]: Fixed.

File: cli/src/diff_util.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (ob) [reply cli/src/diff_util.rs]: I'm going to leave this for now and maybe take a look at what it'd take to allow combining the options.

File: cli/src/diff_util.rs [hunk 8]
```diff
@@ -749,6 +760,108 @@ pub fn show_diff_summary(
...
+                let stat = get_diff_stat(path, &[], &right_content);
+                stats.push(stat);
+            }
+            tree::Diff::Removed(left) => {
+                let left_content = diff_content(workspace_command.repo(), &repo_path, &left)?;
+                let stat = get_diff_stat(path, &left_content, &[]);
+                stats.push(stat);
+            }
+        }
+        max_path_length = max(max_path_length, repo_path.to_internal_file_string().len());
```
Comment (martinvonz) [inline cli/src/diff_util.rs:818]: Use `path` here now that you have it.

File: cli/src/diff_util.rs [hunk 9]
```diff
@@ -749,6 +760,108 @@ pub fn show_diff_summary(
...
+        let mut total_added = 0;
+        let mut total_removed = 0;
+        for stat in &stats {
+            total_added += stat.added;
+            total_removed += stat.removed;
+            let bar_added = (stat.added as f64 * factor).ceil() as usize;
+            let bar_removed = (stat.removed as f64 * factor).ceil() as usize;
+            // pad to max_path_length
+            write!(
+                formatter.labeled("hunk-header"),
```
Comment (martinvonz) [inline cli/src/diff_util.rs:845]: We should probably pick a different (new) label name for this since it's not really a hunk header. Maybe "stat-line"?

File: cli/src/diff_util.rs [hunk 10]
```diff
@@ -749,6 +760,108 @@ pub fn show_diff_summary(
...
+                " {:<max_path_length$} | {:>4} ",
+                stat.path,
+                stat.added + stat.removed,
+            )?;
+            write!(formatter.labeled("added"), "{}", "+".repeat(bar_added))?;
+            writeln!(formatter.labeled("removed"), "{}", "-".repeat(bar_removed))?;
+        }
+        writeln!(
+            formatter.labeled("hunk-header"),
```
Comment (martinvonz) [inline cli/src/diff_util.rs:854]: This is also not a hunk header, so let's use a different name. Maybe "stat-summary"?

File: cli/src/diff_util.rs [hunk 11]
```diff
@@ -36,12 +37,15 @@ use crate::merge_tools::{self, ExternalMergeTool};
...
 
 #[derive(clap::Args, Clone, Debug)]
-#[command(group(clap::ArgGroup::new("short-format").args(&["summary", "types"])))]
+#[command(group(clap::ArgGroup::new("short-format").args(&["summary", "stat", "types"])))]
 #[command(group(clap::ArgGroup::new("long-format").args(&["git", "color_words", "tool"])))]
 pub struct DiffFormatArgs {
     /// For each path, show only whether it was modified, added, or removed
     #[arg(long, short)]
     pub summary: bool,
+    // Show an histogram of the changes
```
[the remark this answers is not in the document]
Comment (ob) [reply cli/src/diff_util.rs]: huh, I've always said that wrong apparently ;-)

File: cli/src/diff_util.rs [hunk 12]
```diff
@@ -749,6 +760,108 @@ pub fn show_diff_summary(
...
+                let stat = get_diff_stat(path, &left_content, &[]);
+                stats.push(stat);
+            }
+        }
+        max_path_length = max(max_path_length, repo_path.to_internal_file_string().len());
+        max_diffs = stats
+            .iter()
+            .map(|stat| stat.added + stat.removed)
+            .max()
+            .unwrap_or(0);
```
[the remark this answers is not in the document]
Comment (ob) [reply cli/src/diff_util.rs]: Thanks for the careful review... I gotta pay more attention before submitting :-)

File: cli/src/diff_util.rs (hunk 9 above)
Comment (ob) [reply cli/src/diff_util.rs]: Done

File: cli/src/diff_util.rs (hunk 10 above)
Comment (ob) [reply cli/src/diff_util.rs]: Done.

File: cli/src/diff_util.rs (hunk 8 above)
Comment (ob) [reply cli/src/diff_util.rs]: Yeah, good idea.

**Comment by martinvonz** [comment]
Thanks for fixing. Will there be a test or two as well? :) They would probably belong in https://github.com/martinvonz/jj/blob/main/cli/tests/test_diff_command.rs

**Comment by ob** [comment]
Ah, thanks for the pointer to where the tests are. I'll figure out a way to add some tests...

**Comment by ob** [comment]
Ok, added some tests (and of course found a new bugs)... if there are other corner cases you can think of let me know and I'll add more tests...

File: cli/src/diff_util.rs [hunk 13]
```diff
@@ -749,6 +760,107 @@ pub fn show_diff_summary(
...
+            // pad to max_path_length
+            write!(
+                formatter.labeled("stat-line"),
+                " {:<max_path_length$} | {:>number_padding$}{}",
+                stat.path,
+                stat.added + stat.removed,
+                if bar_added + bar_removed > 0 { " " } else { "" },
+            )?;
+            write!(formatter.labeled("added"), "{}", "+".repeat(bar_added))?;
+            writeln!(formatter.labeled("removed"), "{}", "-".repeat(bar_removed))?;
```
Comment (martinvonz) [inline cli/src/diff_util.rs:847]: Sorry that I didn't notice this before, but the "stat-line" label should probably apply to the whole line. Alternatively, we can drop the label for now since we're not coloring it differently anyway. I would drop it.

File: cli/src/diff_util.rs [hunk 14]
```diff
@@ -749,6 +760,107 @@ pub fn show_diff_summary(
...
+        let mut total_removed = 0;
+        for stat in &stats {
+            total_added += stat.added;
+            total_removed += stat.removed;
+            let bar_added = (stat.added as f64 * factor).ceil() as usize;
+            let bar_removed = (stat.removed as f64 * factor).ceil() as usize;
+            // pad to max_path_length
+            write!(
+                formatter.labeled("stat-line"),
+                " {:<max_path_length$} | {:>number_padding$}{}",
```
Comment (martinvonz) [inline cli/src/diff_util.rs:841]: Is the leading space intentional?

File: cli/src/diff_util.rs [hunk 15]
```diff
@@ -749,6 +760,107 @@ pub fn show_diff_summary(
     })
 }
 
+struct DiffStat {
+    path: String,
+    added: usize,
+    removed: usize,
+}
+
+fn get_diff_stat(repo_path: String, left_content: &[u8], right_content: &[u8]) -> DiffStat {
```
Comment (martinvonz) [inline cli/src/diff_util.rs:769]: nit: rename `repo_path` to `ui_path` or just `path` since it's no longer a `RepoPath`

File: cli/tests/test_diff_command.rs [hunk 16]
```diff
@@ -106,6 +113,15 @@ fn test_diff_basic() {
     @@ -1,0 +1,1 @@
     +foo
     "###);
+
+    let stdout = test_env.jj_cmd_success(&repo_path, &["diff", "--stat"]);
+    let stdout = strip_lines_prefix(stdout, " ");
```
Comment (martinvonz) [inline cli/tests/test_diff_command.rs:118]: Why strip a prefix? I'd prefer to see the actual output in the test, unless that causes problems for some reason

**Review by martinvonz — APPROVED** [review]
Looks good. Thanks again. Feel free to merge once you've resolved the remaining comments.

File: cli/src/diff_util.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (martinvonz) [reply cli/src/diff_util.rs]: > maybe take a look at what it'd take to allow combining the options.

I *think* it's just a matter of UX - do we want to allow `jj diff --git --color-words` or do we want to tell the user that they probably made a mistake? It might be okay to just drop the restriction.

File: cli/src/diff_util.rs (hunk 13 above)
Comment (ob) [reply cli/src/diff_util.rs:847]: > we're not coloring it differently anyway

I think we are? Isn't that were the green/red colors come from?

image

Comment (martinvonz) [reply cli/src/diff_util.rs:847]: Do you mean the green and red pluses and minuses? That's because of the `added` and `removed` labels, not because of the `stat-line` label, I'm pretty sure.

File: cli/src/diff_util.rs (hunk 14 above)
Comment (ob) [reply cli/src/diff_util.rs:841]: Yeah, I don't know where it comes from but if you look at `git diff --stat` it also adds the leading space. I think it comes from the original `diffstat` program that indents with one space.

```
jj ❯ git diff 2102de94 e414f3b7 | diffstat
 CHANGELOG.md                       |    4 ++
 cli/src/cli_util.rs                |   34 +++++-----------------
 cli/src/commands/mod.rs            |   51 ++++++++++++++++------------------
 cli/src/config-schema.json         |    5 +++
 cli/src/config/colors.toml         |    1
 cli/src/diff_util.rs               |   20 +++++++++----
 cli/src/merge_tools.rs             |   55 ++++++++++++++++++++----------------
 cli/tests/test_describe_command.rs |   89 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-
 cli/tests/test_diff_command.rs     |   23 +++++++++++++++
 docs/config.md                     |    4 +-
 docs/operation-log.md              |    2 -
 lib/src/conflicts.rs               |   17 ++++-------
 lib/src/default_index_store.rs     |  115 ++++++++++++++++++++++-------------------------------------------------------
 lib/src/gitignore.rs               |    7 +---
 lib/src/local_backend.rs           |    4 --
 lib/src/repo.rs                    |    7 ++--
 lib/src/working_copy.rs            |   76 ++++++++++++++++++++++++--------------------------
 lib/src/workspace.rs               |    7 ++--
 lib/tests/test_conflicts.rs        |    8 -----
 19 files changed, 288 insertions(+), 241 deletions(-)
```

File: cli/src/diff_util.rs (hunk 15 above)
Comment (ob) [reply cli/src/diff_util.rs:769]: Ah, good catch!

File: cli/tests/test_diff_command.rs (hunk 16 above)
Comment (ob) [reply cli/tests/test_diff_command.rs:118]: I fiddled with the `r###"` format and could not get it to *not* strip the leading whitespace so I ended up stripping it from the output. If you have a better idea I'm all ears.

File: cli/src/diff_util.rs (hunk 14 above)
Comment (martinvonz) [reply cli/src/diff_util.rs:841]: Weird! Do you think we need to match that weirdness? I'd rather remove it if possible.

File: cli/tests/test_diff_command.rs (hunk 16 above)
Comment (martinvonz) [reply cli/tests/test_diff_command.rs:118]: Ah, I see, so it's related to that other comment I had about a leading space then. Hopefully we can drop that leading space and not have this problem.

File: cli/src/diff_util.rs (hunk 13 above)
Comment (ob) [reply cli/src/diff_util.rs:847]: I thought you meant apply `stat-line` to the whole line? Won't that make the whole thing white? Sorry, I'm not terribly familiar with the formatter...

Comment (ob) [reply cli/src/diff_util.rs]: Ok, I after re-reading I think I understand what you mean. I dropped the labeling part...

File: cli/src/diff_util.rs (hunk 14 above)
Comment (ob) [reply cli/src/diff_util.rs:841]: We can drop it I guess? I've been cargo-cutting it for like a decade :-)

File: cli/src/diff_util.rs (hunk 13 above)
Comment (martinvonz) [reply cli/src/diff_util.rs]: Ah, nope. It's kind of like a very simplified CSS. The labels added here form a tree. Add the "diff" label at the outermost level. Then you add the "added" and "removed" labels around the pluses and minuses. The formatter then uses those labels to choose the colors from here: https://github.com/martinvonz/jj/blob/c3d9ba9ca91e4a4fc415db60af0f0ec9697d0aac/cli/src/config/colors.toml#L58-L59

Looks like we don't have any documentation about that :( We should add that.

File: cli/src/diff_util.rs (hunk 14 above)
Comment (ob) [reply cli/src/diff_util.rs]: ok, dropped.

Comment (martinvonz) [reply cli/src/diff_util.rs]: Yes, let's not waste a column until we find a reason to do it (at which point it's no longer "waste") :)

File: cli/tests/test_diff_command.rs (hunk 16 above)
Comment (ob) [reply cli/tests/test_diff_command.rs]: Done.

Comment (martinvonz) [reply cli/tests/test_diff_command.rs]: Thanks! All looks good to me now. Feel free to mark this resolved and enable auto-merge.
